pkg/didcomm/packer/authcrypt/pack.go

```diff
@@ -0,0 +1,231 @@
+// The returned Packer contains all the information required to pack and unpack payloads.
+func New(ctx packer.Provider, encAlg jose.EncAlg) (*Packer, error) {
+	k := ctx.KMS()
+
+	store, err := ctx.StorageProvider().OpenStore(ThirdPartyKeysDB)
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt: %w", err)
+	}
+
+	return &Packer{
+		kms:    k,
+		encAlg: encAlg,
+		store:  store,
+	}, nil
+}
+
+// Pack will encode the payload argument
+// Using the protocol defined by the Authcrypt message of Aries RFC 0334
+// with the following arguments:
+// payload: the payload message that will be protected
+// senderID: the key id of the sender (stored in the KMS)
+// recipientsPubKeys:
+func (p *Packer) Pack(payload, senderID []byte, recipientsPubKeys [][]byte) ([]byte, error) {
+	if len(recipientsPubKeys) == 0 {
+		return nil, fmt.Errorf("authcrypt Pack: empty recipientsPubKeys")
+	}
+
+	recECKeys, err := unmarshalRecipientKeys(recipientsPubKeys)
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Pack: failed to convert recipient keys: %w", err)
+	}
+
+	kh, err := p.kms.Get(string(senderID))
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Pack: failed to get sender key from KMS: %w", err)
+	}
+
+	jweEncrypter, err := jose.NewJWEEncrypt(p.encAlg, string(senderID), kh.(*keyset.Handle), recECKeys)
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Pack: failed to new JWEEncrypt instance: %w", err)
+	}
+
+	jwe, err := jweEncrypter.Encrypt(payload)
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Pack: failed to encrypt payload: %w", err)
+	}
+
+	var s string
+
+	if len(recipientsPubKeys) == 1 {
+		s, err = jwe.CompactSerialize(json.Marshal)
+	} else {
+		s, err = jwe.FullSerialize(json.Marshal)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Pack: failed to serialize JWE message: %w", err)
+	}
+
+	return []byte(s), nil
+}
+
+func unmarshalRecipientKeys(keys [][]byte) ([]*composite.PublicKey, error) {
+	var pubKeys []*composite.PublicKey
+
+	for _, key := range keys {
+		var ecKey *composite.PublicKey
+
+		err := json.Unmarshal(key, &ecKey)
+		if err != nil {
+			return nil, err
+		}
+
+		pubKeys = append(pubKeys, ecKey)
+	}
+
+	return pubKeys, nil
+}
+
+// Unpack will decode the envelope using a standard format
+func (p *Packer) Unpack(envelope []byte) (*transport.Envelope, error) {
+	jwe, err := jose.Deserialize(string(envelope))
+	if err != nil {
+		return nil, fmt.Errorf("authcrypt Unpack: failed to deserialize JWE message: %w", err)
+	}
+
+	for i := range jwe.Recipients {
+		var (
+			kid                   string
+			kh                    interface{}
+			pt, ecdh1puPubKeyByes []byte
+		)
+
+		kid, err = getKID(i, jwe)
+		if err != nil {
+			return nil, fmt.Errorf("authcrypt Unpack: %w", err)
+		}
+
+		kh, err = p.kms.Get(kid)
+		if err != nil {
+			if strings.EqualFold(err.Error(), fmt.Sprintf("cannot read data for keysetID %s: %s", kid,
+				storage.ErrDataNotFound)) {
+				retriesMsg := ""
```

Comment:
It would be enough if you want to check `storage.ErrDataNotFound` error.
```
if errors.Is(err, storage.ErrDataNotFound) {
```

Reply:
the error is wrapped, can't use errors.Is(), unless the wrapping message is removed.

Reply:
this is exactly what errors.Is() does. From the doc: `Is reports whether any error in err's chain matches target.`

Reply:
I have tried it and also thought the same thing, unfortunately `err.Is(storage.ErrDataNotFound)` returns false

Reply:
seems to work.. I must have tried something else.. will update..

Reply:
updated